core/src/main/scala/kafka/utils/Throttler.scala

```diff
@@ -36,7 +36,7 @@ import scala.math._
  * @param time: The time implementation to use
  */
 @threadsafe
-class Throttler(desiredRatePerSec: Double,
+class Throttler(var desiredRatePerSec: Double,
```

Comment:
If the `desiredRatePerSec` can be changed by other threads, should we `volatile` it?

Reply:
@showuon given that we're synchronizing on `lock` anyway at the point of use, maybe synchronizing in `updateDesiredRatePerSec()` would actually be better?

Reply:
@tombentley Good point! +1. Thanks.

Reply:
Thanks for pointing out, @showuon and @tombentley.
Indeed, it needs volatile or synchronized.

I would like your opinion. Should it use volatile? or synchronized?
I have changed the implementation once with volatile, so if you prefer synchronized, please let me know. https://github.com/apache/kafka/pull/12296/commits/9b2477dc8a38db7e41281a6c5aba9c19348e8d5d

If use synchronized, do we apply a synchronized block to this object as following code?
I think `lock` object cannot be used for synchronized in updateDesiredRatePerSec() like `lock synchronized {…} block in maybeThrottle()`, it may take more than a few minute due to frequent sleeps depending on Throttler.desiredRatePerSec, and may block reconfigure() thread at updateDesiredRatePerSec() as well. 

```
  def updateDesiredRatePerSec(updatedDesiredRatePerSec: Double): Unit = {
    this synchronized {
      desiredRatePerSec = updatedDesiredRatePerSec;
    }
  }

```

Reply:
Thanks for pointing it out. Yes, it sleeps inside the lock. At first, I thought it is a but, but after further thought, I think it is intended because there might be multiple threads try to call `maybeThrottle` method. We'd like to calculate the throttle time sequentially (I guess). This code is written since 2011, so I think we just keep it as is. 
In that case, I agree `volatile` is a better solution. Otherwise, the reconfigure change might not be able to take effect immediately.

@tombentley , WDYT?

Reply:
That seems reasonable.